# 03 Modelling

## 1. Scoring harness and mean-prediction benchmark

The competition metric is hidden. `inrange.scoring` implements our own approximation of it, and every evaluation in this notebook reports the composite **and** each component, so no conclusion rests on our weights alone.

**Components** (mean over shots): landing position error and apex position error (Euclidean distance over x, y, z, in metres), apex time and landing time error (absolute, seconds), launch spin error (absolute, rpm).

**Scaling.** Each component is divided by the error of predicting the training mean for every shot, computed in-sample on all 491 training rows. A scaled component of 1.0 means "no better than the training mean".

**Weights.** landing position 0.40, apex position 0.25, apex time 0.125, landing time 0.125, spin 0.10. They sum to 1 and follow the stated priority order. They are a judgement call, not the real weights.

**Cross-validation.** Two strategies, always reported together:

* *Session-grouped*: `GroupKFold` (5 folds) on the 11 practice sessions from `01_eda.ipynb`. Whole sessions are held out, which is harder than the real test.
* *Within-session*: 10 repeated random holdouts in which every session holds out the same share of its training rows as the real split put in test (38% to 62%). This mirrors how the real test set was drawn.

Each report also breaks errors down by ball speed band and gives a **test speed mix** composite: the band results reweighted to the share of test shots in each band, because test has far more shots above 70 m/s than train.

In [1]:
import pandas as pd

from inrange.io import TARGET_COLS, load_sample_submission, load_test, load_train
from inrange.features import session_labels, speed_band
from inrange.scoring import (
    COMPONENTS, REFERENCE_SCALES, WEIGHTS, compute_reference_scales, evaluate, report_table, score,
)

pd.set_option("display.width", 200, "display.max_columns", 30, "display.precision", 3)

train = load_train()
test = load_test()
print("weights:", WEIGHTS)
print("reference scales:", {k: round(v, 4) for k, v in REFERENCE_SCALES.items()})
print("scales recomputed from train match:", compute_reference_scales(train) == REFERENCE_SCALES)

weights: {'landing_pos': 0.4, 'apex_pos': 0.25, 'apex_t': 0.125, 'landing_t': 0.125, 'spin': 0.1}
reference scales: {'landing_pos': 42.0939, 'apex_pos': 31.3786, 'apex_t': 0.4651, 'landing_t': 0.8568, 'spin': 2036.4877}
scales recomputed from train match: True


In [2]:
bands = pd.DataFrame({
    "train": speed_band(train).value_counts(),
    "test": speed_band(test).value_counts(),
}).reindex(["<50", "50-70", ">=70"])
bands["train share"] = bands["train"] / bands["train"].sum()
bands["test share"] = bands["test"] / bands["test"].sum()
print(bands)

train_sessions, test_sessions = session_labels(train, test)
print(pd.DataFrame({"train": train_sessions.value_counts(), "test": test_sessions.value_counts()}).sort_index().T)

            train  test  train share  test share
speed_band                                      
<50           165   179        0.336       0.320
50-70         289   271        0.589       0.485
>=70           37   109        0.075       0.195
session  0   1   2   3   4   5   6   7   8   9   10
train    12  45  61  26  21  88  72  43  29  38  56
test     20  61  60  20  29  93  93  66  27  55  35


### Benchmark

Two versions of the mean prediction:

* **sample_submission constants.** The fixed values from `sample_submission.csv`. These equal the full training mean, so they have seen every validation fold's targets. Scored in-sample, this gives exactly 1.0 on every component.
* **fold-train mean.** The mean of each fold's training rows only. This is the honest out-of-fold version and the number later models must beat.

In [3]:
sample_constants = load_sample_submission()[TARGET_COLS].iloc[0]

def predict_sample_constants(train_rows, val_rows):
    return pd.DataFrame([sample_constants] * len(val_rows), index=val_rows.index)

def predict_fold_mean(train_rows, val_rows):
    return pd.DataFrame([train_rows[TARGET_COLS].mean()] * len(val_rows), index=val_rows.index)

in_sample = score(train[TARGET_COLS], predict_sample_constants(train, train))
print("in-sample sample_submission composite:", round(in_sample["composite"], 6))

show = ["strategy", "n", "composite"] + COMPONENTS + [f"{c}_scaled" for c in COMPONENTS]
benchmarks = {}
for name, fit_predict in [("sample_submission constants", predict_sample_constants),
                          ("fold-train mean", predict_fold_mean)]:
    benchmarks[name] = evaluate(fit_predict, train, test)
    print(f"\n{name}")
    print(report_table(benchmarks[name])[show])

in-sample sample_submission composite: 1.0



sample_submission constants
                       strategy       n  composite  landing_pos  apex_pos  apex_t  landing_t      spin  landing_pos_scaled  apex_pos_scaled  apex_t_scaled  landing_t_scaled  spin_scaled
overall         session_grouped   491.0      1.000       42.094    31.379   0.465      0.857  2036.488               1.000            1.000          1.000             1.000        1.000
test speed mix  session_grouped   491.0      1.101       46.893    35.223   0.512      0.924  2078.037               1.114            1.123          1.101             1.079        1.020
band <50 m/s    session_grouped   165.0      1.154       49.554    37.312   0.512      0.895  2399.560               1.177            1.189          1.101             1.045        1.178
band 50-70 m/s  session_grouped   289.0      0.821       33.524    24.537   0.395      0.772  1804.365               0.796            0.782          0.850             0.900        0.886
band >=70 m/s   session_grouped    37.0  


fold-train mean
                       strategy       n  composite  landing_pos  apex_pos  apex_t  landing_t      spin  landing_pos_scaled  apex_pos_scaled  apex_t_scaled  landing_t_scaled  spin_scaled
overall         session_grouped   491.0      1.006       42.283    31.538   0.465      0.857  2088.635               1.004            1.005          1.000             1.000        1.026
test speed mix  session_grouped   491.0      1.107       47.126    35.406   0.513      0.926  2123.763               1.120            1.128          1.102             1.080        1.043
band <50 m/s    session_grouped   165.0      1.159       49.697    37.445   0.513      0.895  2446.641               1.181            1.193          1.102             1.045        1.201
band 50-70 m/s  session_grouped   289.0      0.827       33.695    24.687   0.394      0.771  1865.377               0.800            0.787          0.848             0.900        0.916
band >=70 m/s   session_grouped    37.0      1.718   

In [4]:
honest = benchmarks["fold-train mean"]
print("session-grouped, per fold:")
print(honest["session_grouped"].by_fold()[["n", "composite", "landing_pos", "apex_pos", "spin"]])
repeats = honest["within_session"].by_fold()["composite"]
print("\nwithin-session, composite per repeat:", repeats.round(3).tolist())
print(f"spread across repeats: min {repeats.min():.3f}, max {repeats.max():.3f}, sd {repeats.std():.3f}")

session-grouped, per fold:
          n  composite  landing_pos  apex_pos      spin
fold                                                   
0     109.0      0.875       38.382    27.204  2008.164
1      98.0      0.993       40.426    30.842  2328.550
2      90.0      1.135       48.161    35.828  1999.552
3      94.0      1.004       41.495    31.438  1835.391
4     100.0      1.046       43.807    33.177  2259.455

within-session, composite per repeat: [1.037, 0.997, 1.027, 1.013, 0.993, 0.993, 1.007, 1.029, 1.028, 1.045]
spread across repeats: min 0.993, max 1.045, sd 0.019


### Reading the benchmark

* The honest mean prediction scores about **1.01** under both strategies. Grouped and within-session CV agree here because a constant cannot exploit session information; they will diverge once models use session-level signal.
* Reweighting to the test speed mix raises the composite to about **1.11**. The fast band (70 m/s and above) scores about 1.7 because fast shots sit far from the mean, and test has about 2.6 times the share of those shots that train has. Any model's overall CV score will flatter its likely test score in the same way, so the test-mix row is the better guide.
* Only 37 training shots fall in the fast band, so fast-band numbers are noisy.
* Within-session repeats vary by a few hundredths in composite, so differences between models smaller than about 0.02 should not be trusted without more repeats.
* **Scale caveat.** Mean-based scaling makes every component exactly 1.0 for the mean prediction, so no component dominates *here* by construction. Once a model exists this will change: landing and apex positions are highly predictable from ball speed, while spin and times are harder. A model could cut landing error to a small fraction of its scale while spin stays near 1.0, at which point spin (weight 0.10) may contribute as much to the composite as landing (weight 0.40). Watch the scaled columns, not just the composite.